standards/containerisation_standards.md

```diff
@@ -0,0 +1,32 @@
+# Containerisation
+
+## We use Docker containers for delivery of all bespoke software being deployed to AWS and Azure
+Where we develop bespoke software, we default to using a Docker container as the deployable item.
+
+Unless the technology does not support containerisation, all deviations from this standard will be treated as architectural exceptions.
+
+### Rationale
+Containerisation is _most_ applicable to certain specific scenarios, such as highly-available applications with varying demand or when using technologies that have a frequent update cycle.
+
+However, running and managing a delivery pipeline and operational environment that supports _both_ containers and more traditional virtual machines results in a duplication of effort and technical solutions.
+
+Therefore, we will seek to minimise or eliminate this duplication by _always_ preferring containerisation.
+
+Containerisation also allows us to better manage a diverse application estate as it somewhat mitigates the need for across the board standardisation and large, coordinated upgrade projects, allowing instead for more nuanced decision-making.
+
+## We use AWS Lambda and Azure Functions where it is justified
+There are cases where it will make sense to use a serverless platform rather than a containerised solution, but these _must_ be individually justified and recorded as formal architectural decisions.
+
+Additionally, justifications for using serverless platforms can _only_ be considered for Javascript or C# functions.
+
+### Rationale
+Serverless platforms provide a greatly reduced operational burden, but are not so well represented in development and have certain limitations that can make their use inappropriate.
+They also represent some vendor lock-in that, whilst minimal, is still not preferred.
```

Comment:
Vendor lock in is a big negative to a serverless approach.  I appreciate that it's probably not much effort to rewrite an Azure function to an AWS equivalent if migration to another cloud service provided is required.  But it becomes an additionally complicated if your DR is to failover to another provider, you'd need to ensure that you'd already done replicated everything twice.

Another complication is if you're running a federated Kubernetes cluster across Azure and AWS, by adding serverless functions you're creating vendor lock in somewhere that destroys the high availability of your Kubernetes cluster.

Reply:
I agree - that's why containers are the preferred/default option but serverless is identified as an acceptable alternative but only where its use is explicitly justified.

The "minimal" vendor lock-in is really in comparison to things like Logic Apps where the lock-in is absolute.  There's always a small amount of lock-in with running on Cloud, even if it's just something in the deployment scripts or monitoring logs or something like that.  So serverless felt like like it was delivering benefit without adding too much lock-in.

This is another area where further, more detailed, standards will be required, but I felt it was probably better to get _something_ written down and agreed first.